### PR TITLE
Crawler for Externally Orchestrated Jobs with Failing Configuration

### LINKED ISSUE #266: Assessment for `RunSubmit` API usages
**Problem statement**

Existing processes that create `RunSubmit` Jobs (the so-called ephemeral jobs) with the following properties:

- DBR > 11.X
- [AWS-specific] Uses `instance_profile` 
- Has no `data_security_mode` property specified in the job creation request
- [Azure-specific] some configurations in the `spark_conf` (To be confirmed)

May be broken when UC is enabled in a given workspace (meaning when workspace is assigned to the metastore). 

**Why this happens?**

When UC is enabled, all `RunSubmit` Jobs with DBR 11.X+ will by default UC-enabled. If there is a conflict in permissions between UC and service principal (e.g. instance profile), the job will fail. 

The reason and the change is described [here](https://docs.databricks.com/en/archive/runtime-release-notes/11.1.html#job-clusters-default-to-single-user-access-mode-with-databricks-runtime-111-and-higher). 

**How can we identify the identical `RunSubmit`?**

This requires additional internal discussion.

# TODO:

- [ ] get a list of all (persisted) jobs
- [ ] get a list of all job runs
- [ ] find job runs that have no persisted job (out of workflow)
- [ ] group all job runs from non-persisted jobs to approximate the number of unique airflow/azure data factory DAGs.
- [ ] identify job runs that do not include data_security_mode in the job creation request and are run against 11.x  compute

### PR DESCRIPTION
**Pull request by zpappa**
Resolves #266

**Added ExternallyOrchestratedJobsWithFailingConfigCrawler**
Added a crawler to look at JobRuns from the SDK and determine which of the job runs are from the RunsSubmit API

**Added Unit Tests**
Added tests to cover basic logic and some edge cases

**Integration Tests Pending**

### CODE REVIEW

**Review by nfx — CHANGES_REQUESTED** [review]
needs a passing integration test. please attach a screenshot once you get it to work locally

File: src/databricks/labs/ucx/assessment/crawlers.py [hunk 1]
```diff
@@ -57,6 +61,14 @@ class PipelineInfo:
     failures: str
 
 
+@dataclass
```
Comment (FastLee) [inline src/databricks/labs/ucx/assessment/crawlers.py:64]: Do we want to capture only the failing ones or all of them with the failing one anotated?

File: src/databricks/labs/ucx/assessment/crawlers.py [hunk 2]
```diff
@@ -65,6 +77,29 @@ def _azure_sp_conf_present_check(config: dict) -> bool:
     return False
 
 
+def is_custom_image(version_string: str):
+    """
+    Is this a custom version?
+    """
```
Comment (FastLee) [inline src/databricks/labs/ucx/assessment/crawlers.py:83]: Does it require implementation?

File: src/databricks/labs/ucx/assessment/crawlers.py [hunk 3]
```diff
@@ -65,6 +77,29 @@ def _azure_sp_conf_present_check(config: dict) -> bool:
...
+    Is left version greater than right version
+    """
+    if left_version is None or is_custom_image(left_version):
+        return False
+    if right_version is None or is_custom_image(right_version):
+        return True
+    pattern = r"^(?P<major>\d+)?\.(?P<minor>\d+)?\.(?P<patch>[\dx]+)?.*"
+    lvg = re.match(pattern, left_version)
+    rvg = re.match(pattern, right_version)
+    left = (int(lvg.group("major")), int(lvg.group("minor")))
```
Comment (FastLee) [inline src/databricks/labs/ucx/assessment/crawlers.py:98]: Do we have a proper unit test for that?

File: src/databricks/labs/ucx/assessment/crawlers.py [hunk 4]
```diff
@@ -84,6 +119,26 @@ def spark_version_compatibility(spark_version: str) -> str:
     return "supported"
 
 
+def get_job_cluster_from_task(
+    task: RunTask, job_run: BaseRun, all_clusters: dict[str, ClusterDetails]
```
Comment (FastLee) [inline src/databricks/labs/ucx/assessment/crawlers.py:123]: Re factor job cluster to use the same mechanism

File: src/databricks/labs/ucx/assessment/crawlers.py [hunk 5]
```diff
@@ -250,3 +305,142 @@ def snapshot(self) -> list[ClusterInfo]:
...
+            sbe,
+            "hive_metastore",
+            schema,
+            "ext_orc_job_runs_with_failing_config",
+            ExternallyOrchestratedJobRunWithFailingConfiguration,
+        )
+        self._ws = ws
+
+    def _crawl(self) -> list[ExternallyOrchestratedJobRunWithFailingConfiguration]:
+        no_of_days_back = datetime.timedelta(days=30)  # todo make configurable in yaml?
```
Comment (FastLee) [inline src/databricks/labs/ucx/assessment/crawlers.py:332]: Make timedelta externally configurable

**Review by FastLee — CHANGES_REQUESTED** [review]
